File: android/src/main/java/com/gt/plugin/background/geolocation/Trip.java
package com.gt.plugin.background.geolocation;

import android.location.Location;
import android.util.Log;

import com.getcapacitor.JSObject;

import java.util.ArrayList;
import java.util.List;

public class Trip {

    static final String TAG = "Trip";

    List<Location> puntos;
    Double distancia;
    Long timestampInicio;
    Long timestampFin;

    public Trip() {
        this.puntos = new ArrayList<>();
        this.distancia = 0.0;
        this.timestampInicio = System.currentTimeMillis();
    }

    public static JSObject toJSObject(Trip trip) {
        return toJSObject(trip, false);

    }

    public static JSObject toJSObject(Trip trip, boolean withPath) {
        JSObject ret = null;
        if (trip != null) {
            ret = new JSObject();
            ret.put("distancia", trip.getDistancia());
            ret.put("timestampInicio", trip.getTimestampInicio());

            ret.put("timestampFin", trip.getTimestampFin());
            if (withPath) {
                List<JSObject> puntos = new ArrayList<>();
                for (Location punto : trip.getPuntos()) {
                    puntos.add(LocationPayloadBuilder.locationToJson(punto));
                }
                ret.put("puntos", puntos);
            }
        }
        return ret;
    }


    public List<Location> getPuntos() {
        return puntos;
    }

    public Double getDistancia() {
        return distancia;
    }

    public Long getTimestampInicio() {
        return timestampInicio;
    }

    public Long getTimestampFin() {
        return timestampFin;
    }

    public void setTimestampFin(Long timestampFin) {
        this.timestampFin = timestampFin;
    }

    public void addPunto(Location location) {

        if (location == null) {
            return;
        }

        Location ultimo = this.getUltimoPunto();

        if (ultimo != null) {
            float curDist = ultimo.distanceTo(location);
            if (curDist > 30f) {
                this.distancia += curDist;
                puntos.add(location);
            } else {
                Log.d(TAG, "Punto ignorado: " + curDist);
            }
        } else {
            puntos.add(location);
        }
    }

    public Location getUltimoPunto() {
        if (puntos.isEmpty()) {
            return null;
        }

        return puntos.get(puntos.size() - 1);
    }
}

File: ios/Sources/GtBackgroundGeolocationPlugin/Trip.swift
import Foundation
import CoreLocation

public class Trip {
    private static let minPointDistance: Double = 30.0

    public var puntos: [CLLocation]
    public var distancia: Double
    public var timestampInicio: Int64
    public var timestampFin: Int64?

    public init() {
        self.puntos = []
        self.distancia = 0.0
        self.timestampInicio = Int64(Date().timeIntervalSince1970 * 1000)
        self.timestampFin = nil
    }

    public func addPunto(_ location: CLLocation) {
        if let ultimo = puntos.last {
            let curDist = location.distance(from: ultimo)
            if curDist > Trip.minPointDistance {
                distancia += curDist
                puntos.append(location)
            }
        } else {
            puntos.append(location)
        }
    }

    public func toDictionary(withPath: Bool = false) -> [String: Any] {
        var dict: [String: Any] = [
            "distancia": distancia,
            "timestampInicio": timestampInicio
        ]
        if let timestampFin = timestampFin {
            dict["timestampFin"] = timestampFin
        }
        if withPath {
            dict["puntos"] = puntos.map { LocationPayloadBuilder.locationToDictionary($0) }
        }
        return dict
    }
}
